### PR TITLE
[WIP][TF] Fix t5 embeddings

### LINKED ISSUE #13839: TF mT5 model is not adding new tokens into it's vocabulary.
## Environment info

- `transformers` version: 4.11.2
- Platform: google colab
- Python version: 3.7.12
- PyTorch version (GPU?):
- Tensorflow version (GPU?): 2.6.0
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

## Information

Model I am using (Bert, XLNet ...):
I am using mT5 model. The problem arised when I am trying to resize the token embeddings. I added new special tokens to my tokenizer's vocabulary, when I resized the model's token embeddings according to the length of the tokenizer, I am facing the issue.  

Link for the notebook:  https://colab.research.google.com/drive/1ooKa5aQ_FAEnicxBL8bJnNAO4H9vFuv-?usp=sharing
Code from the notebook:  
!pip install transformers
!pip install sentencepiece
from transformers import TFMT5ForConditionalGeneration, T5Tokenizer
model = TFMT5ForConditionalGeneration.from_pretrained("google/mt5-base")
tokenizer = T5Tokenizer.from_pretrained("google/mt5-base")
tokenizer.add_special_tokens({'bos_token':'<bos>','eos_token':'<eos>'})
model._resize_token_embeddings(len(tokenizer))

The error message is: 
ValueError: Attempt to convert a value (None) with an unsupported type (<class 'NoneType'>) to a Tensor.

The problem arises when using:

* [ ] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [ ] my own task or dataset: (give details below)

## To reproduce

1. You can directly run the code & check the error coming from the last line.
2. Also you can go the colab notebook & have a look at the error as well.
3.

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR is a first attempt to fix: #13839. In short T5 models that don't have input and output embeddings tied, can resize the embeddings.

Overall this whole TF resize embedding layer logic is incredibly complex and not readable...IMO, we should do a bigger refactor here.

### TODO:

- [ ] Add test

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/modeling_tf_utils.py [hunk 1]
```diff
@@ -1116,14 +1120,15 @@ def _get_resized_lm_head_decoder(self, old_lm_head_decoder, new_num_tokens):
             self._get_word_embedding_weight(self.get_input_embeddings()) == old_lm_head_decoder
         )
 
-        if old_lm_head_decoder is not None and not is_input_output_equals:
+        if old_lm_head_decoder is not None and (not is_input_output_equals or not self.config.tie_word_embeddings):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:1123]: I don't think `is_input_output_equals` is enough to decide whether the input and outputs are tied or not -> it's mostly the `tie_word_embeddings` attribute that does that

File: src/transformers/modeling_tf_utils.py [hunk 2]
```diff
@@ -1116,14 +1120,15 @@ def _get_resized_lm_head_decoder(self, old_lm_head_decoder, new_num_tokens):
...
             old_embedding_dim = shape_list(old_lm_head_decoder)[1]
             decoder_mask, current_decoder = init_copy_embeddings(old_lm_head_decoder, new_num_tokens)
+            name = old_lm_head_decoder.name.split(":")[0] if not tf.executing_eagerly() else None
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:1126]: Can't pass name in eager mode so disable it

File: src/transformers/modeling_tf_utils.py [hunk 3]
```diff
@@ -1017,6 +1018,10 @@ def _get_word_embedding_weight(model, embedding_layer):
         if embeds is not None:
             return embeds
 
+        # if embedding_layer is already a `tf.Tensor` simply output it
+        if isinstance(embedding_layer, tf.Tensor):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:1022]: embedding_layer can be a tensor for T5 I think -> so return just return this?

File: src/transformers/modeling_tf_utils.py [hunk 4]
```diff
@@ -610,22 +610,23 @@ def init_copy_embeddings(old_embeddings, new_num_tokens):
     old_num_tokens, old_embedding_dim = shape_list(old_embeddings)
     size_diff = new_num_tokens - old_num_tokens
 
+    # get detached tensor if not already detached
+    old_embeddings = old_embeddings.values() if hasattr(old_embeddings, "values") else old_embeddings
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:614]:
```suggestion
    old_embeddings = old_embeddings.values() if hasattr(old_embeddings, "values") else old_embeddings
```
It's `value` beneath (without an s). Is the s intended?

**Review by sgugger — APPROVED** [review]
Thanks for fixing! Is the method `value` or `values`? The two are used in the diff.

File: src/transformers/modeling_tf_utils.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_tf_utils.py:614]: Good catch! It's `value()` for sure!

**Comment by patrickvonplaten** [comment]
Superseeded by: https://github.com/huggingface/transformers/pull/15567

**Pull request closed without merging**